components/plurals/src/provider.rs

```diff
@@ -368,115 +377,705 @@ pub enum PluralElementsKeysV1 {
+        let default = unsafe { V::from_byte_slice_unchecked(unpacked_bytes.v_bytes) };
+        #[allow(clippy::manual_map)] // more explicit with the unsafe code
+        let specials = if let Some(specials_bytes) = unpacked_bytes.specials_bytes {
+            // Safety: the bytes are valid by invariant
+            Some(unsafe {
+                PluralElementsTupleSliceVarULE::<V>::from_byte_slice_unchecked(specials_bytes)
+            })
+        } else {
+            None
+        };
+        PluralElementsUnpacked {
+            default: (metadata, default),
+            specials,
+        }
+    }
+
+    /// Returns the value for the given [`PluralOperands`] and [`PluralRules`].
+    pub fn get<'a>(&'a self, op: PluralOperands, rules: &PluralRules) -> (FourBitMetadata, &'a V) {
+        let parts = self.as_parts();
+
         let category = rules.category_for(op);
 
-        if op.is_exactly_zero() {
-            if let Some(value) = self
-                .specials
-                .iter()
-                .filter_map(|ule| {
-                    (ule.0 == PluralElementsKeysV1::ExplicitZero.to_unaligned()).then_some(&ule.1)
-                })
-                .next()
-            {
-                return value;
+        match parts.specials {
+            Some(specials) => {
+                if op.is_exactly_zero() {
+                    if let Some(value) = get_special(specials, PluralElementsKeysV1::ExplicitZero) {
+                        return value;
+                    }
+                }
+                if op.is_exactly_one() {
+                    if let Some(value) = get_special(specials, PluralElementsKeysV1::ExplicitOne) {
+                        return value;
+                    }
+                }
+                match category {
+                    PluralCategory::Zero => Some(PluralElementsKeysV1::Zero),
+                    PluralCategory::One => Some(PluralElementsKeysV1::One),
+                    PluralCategory::Two => Some(PluralElementsKeysV1::Two),
+                    PluralCategory::Few => Some(PluralElementsKeysV1::Few),
+                    PluralCategory::Many => Some(PluralElementsKeysV1::Many),
+                    PluralCategory::Other => None,
+                }
+                .and_then(|key| get_special(specials, key))
             }
+            None => None,
         }
+        .unwrap_or(parts.default)
+    }
+}
 
-        if op.is_exactly_one() {
-            if let Some(value) = self
-                .specials
-                .iter()
-                .filter_map(|ule| {
-                    (ule.0 == PluralElementsKeysV1::ExplicitZero.to_unaligned()).then_some(&ule.1)
-                })
-                .next()
-            {
-                return value;
+impl<V> PluralElementsPackedULE<V>
+where
+    V: VarULE + PartialEq + ?Sized,
+{
+    #[cfg(feature = "datagen")]
+    fn to_plural_elements(&self) -> PluralElements<(FourBitMetadata, &V)> {
```

Comment:
this is also only used by a single call site, `serde::Serialize`. I'd prefer if that defined its own local struct with a `derive(Serialize)`, so the `PluralElements` doesn't have to expose `Serialize` (it's not great having a public type be `Serialize` if it doesn't work for non-human readable formats).

Reply:
`PluralElements` _does_ serialize to non-human-readable formats, it's just not optimized for it. I'm convinced; I'll make a private struct for the serialization.